[temp.deduct.general] Clarify explicit-specifier and exception specifications CWG2618

jensmaurer opened the pull request:
Fixes #5129

jensmaurer commented:
@cmeerw, something like this?

cmeerw commented:
yes, looks good to me.

xmh0511 commented:
@jensmaurer  Should we also improve the second sentence of [temp.deduct.general] p8 to make [temp.deduct.general.note.4] have a normal explanation?  The current sentence is that
> Only invalid types and expressions in the immediate context of the function type, its template parameter types, and its explicit-specifier can result in a deduction failure.

I think we may change it to 
> Only invalid types and expressions in the immediate context of the function type **outside of the exception specification**, its template parameter types, and its explicit-specifier can result in a deduction failure.  

In other words, the substitution in exception specification that would result in invalid types and expressions can make the program be ill-formed, which is what the note saying

jensmaurer commented:
@xmh0511, since this is not the place where we specify exactly where we substitute, we can be a bit hazy here. We should not be incorrect. I've fixed the obvious problem, and I've moved the "only" to apply to "immediate context", not to "invalid stuff".

xmh0511 commented on source/templates.tex:
```diff
@@ -7145,9 +7148,9 @@
-its template parameter types,
+Invalid types and expressions only in the immediate context of
+the function's type outside of the exception specification,
+its template parameters,
```
@jensmaurer  Could we change `template parameters` to *template-parameters*; It would indicate the deduction failures in default template arguments are also covered by these rules.

tkoeppe commented:
@jensmaurer Could this be something you could quickly look at in CWG? No objections otherwise, but if you have a slot for GitHub reviews, it would be nice to get core sign-off.

jensmaurer commented:
Created CWG2618 for CWG review.

frederick-vs-ja commented:
Should have been superseded by #5984 (https://github.com/cplusplus/draft/pull/5984/commits/bb4acc841a4b5fbca8b00fc121cc887877583663).
